app/utils/callback/lava.py

```diff
@@ -381,225 +323,289 @@ def _store_lava_json(job_data, meta, base_path=utils.BASE_PATH):
+                if end_line < len(self.log) else end_line
+
+    @property
+    def start_log_line(self):
+        line_number = self._get_log_line_number(self.LOGIN_CASE_END_PATTERN)
+        return 0 if line_number is None else line_number
+
+    def _get_log_line_number(self, pattern):
+        for line_number, line in enumerate(self.log):
+            msg = line.get('msg', '')
+            if pattern.match(unicode(msg)) is not None:
+                return line_number
+
+    def _find_new_end_line(self, end_line_number, test_case_id):
+        signal_snd_template = r'<LAVA_SIGNAL_TESTCASE TEST_CASE_ID={}'
+        signal_snd_text = signal_snd_template.format(test_case_id)
+        signal_snd_pattern = re.compile(signal_snd_text)
+        for i, log_line in enumerate(reversed(self.log[:end_line_number]), 1):
+            if signal_snd_pattern.match(unicode(log_line['msg'])):
+                return end_line_number - i
+        return end_line_number
+
+    def _adjust_log_end_line(self, end_line_number):
+        log_end_line = self.log[end_line_number]
+        test_case_rcvd = \
+            self.SIGNAL_RECEIVED_PATTERN.match(log_end_line['msg'])
+        test_case_id = test_case_rcvd.group(1) if \
+            test_case_rcvd is not None else None
+        if test_case_id:
+            return self._find_new_end_line(end_line_number, test_case_id)
+        return end_line_number
+
+    @staticmethod
+    def _prepare_lines_map(end_lines_map, start_log_line):
+        lines_map = OrderedDict(sorted(end_lines_map.items(),
+                                       key=lambda i: i[1]))
+        start_line = start_log_line
+        for path, end_line in lines_map.items():
+            lines_map[path] = (start_line, end_line)
+            start_line = end_line + 1
+        return lines_map
+
+    def _test_case_iter(self):
+        for group in self.groups:
+            stack = [group]
+            path = [group.get('name')]
+            while stack:
+                node = stack.pop()
+                if node is not group:
+                    path.append(node.get('name'))
+                for test_case in node.get('test_cases', []):
+                    path.append(test_case.get('name'))
+                    yield tuple(path), test_case
+                    path.pop()
+                if node is not group:
+                    path.pop()
+                for sub_group in node.get('sub_groups', []):
+                    stack.append(sub_group)
+
+    def _get_log_lines(self, start_line, end_line):
+        lines = [
+            {
+                'dt': dparser.parse(line['dt']),
+                'msg': line['msg'],
+                'lvl': line['lvl']
+            }
+            for line in self.log[start_line:end_line]
+        ]
+        return lines
+
+    @staticmethod
+    def _filter_log_data(log, filters_funcs):
+        for filter_func in filters_funcs:
+            log[:] = filter(filter_func, log)
+
+
+class LavaPlan(object):
```

Comment:
Would be good to have docstrings in this class.

Reply:
Added docstring